prov/efa/test/efa_unit_test_mocks.c

```diff
@@ -5,68 +5,128 @@
 #include <setjmp.h>
 #include <cmocka.h>
 #include "efa.h"
+#include "efa_unit_test_mocks.h"
 
-void __real_rxr_pkt_handle_send_completion(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry);
+/* mock of rdma-core functions */
+uint8_t g_ibv_gid_mock[EFA_GID_LEN] = {0xfe, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x10, 0xcb, 0x06, 0xff, 0xfe, 0x66, 0x83, 0x2f};
 
-void __wrap_rxr_pkt_handle_send_completion(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+int g_ibv_create_ah_call_counter = 0;
+struct ibv_ah *efa_mock_ibv_create_ah_increase_call_counter(struct ibv_pd *pd, struct ibv_ah_attr *attr)
 {
-	check_expected(ep);
-	check_expected(pkt_entry);
+	g_ibv_create_ah_call_counter += 1;
+	return  __real_ibv_create_ah(pd, attr);
+}
 
-	if ((int)mock() == 4242) {
-		__real_rxr_pkt_handle_send_completion(ep, pkt_entry);
-	}
+int efa_mock_efadv_query_device_return_mock(struct ibv_context *ibv_ctx,
+					    struct efadv_device_attr *attr,
+					    uint32_t inlen)
+{
+	return mock();
 }
 
-void __real_rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
-										   struct rxr_pkt_entry *pkt_entry,
-										   enum rxr_lower_ep_type lower_ep_type);
+struct ibv_send_wr *g_ibv_send_wr_list_head;
+struct ibv_send_wr *g_ibv_send_wr_list_tail;
 
-void __wrap_rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
-										   struct rxr_pkt_entry *pkt_entry,
-										   enum rxr_lower_ep_type lower_ep_type)
+int efa_mock_ibv_post_send_save_send_wr(struct ibv_qp *qp, struct ibv_send_wr *wr,
+					struct ibv_send_wr **bad_wr)
 {
-	check_expected(ep);
-	check_expected(pkt_entry);
-	check_expected(lower_ep_type);
+	struct ibv_send_wr *head, *tail, *entry;
+
+	head = tail = NULL;
+
+	while(wr) {
+		entry = calloc(sizeof(struct ibv_send_wr), 1);
+		if (!entry) {
+			*bad_wr = wr;
+			return ENOMEM;
+		}
 
-	if ((int)mock() == 4242) {
-		__real_rxr_pkt_handle_recv_completion(ep, pkt_entry, lower_ep_type);
+		memcpy(entry, wr, sizeof(struct ibv_send_wr));
+		if (!head)
+			head = entry;
+
+		if (tail)
+			tail->next = entry;
+		tail = entry;
```

Comment:
Not understanding this while loop... did you miss `else {...}` on this line?

Reply:
No. This loop is adding a linked list to an existing list.

`tail` is the tail of the existing list.

if tail is not NULL, then it is pointing to current tail. we need to update it so it is pointing to the new entry.

then we will always make `tail` to point to the new entry.

Reply:
I'm still not getting it:
```c
		if (tail)
			tail->next = entry;

		tail = entry;
```
It reads to me that, if `tail` is set, we will have `tail = tail->next = entry`. Do you mean `tail->next = head` instead?